src/gui/editorwidgets/qgsrelationreferencewidget.cpp

```diff
@@ -155,6 +156,38 @@ QgsRelationReferenceWidget::~QgsRelationReferenceWidget()
     delete mMapTool;
 }
 
+void QgsRelationReferenceWidget::updateIndex()
+{
+  if ( mChainFilters && mComboBox->count() > 0 )
+  {
+    int index = -1;
+
+    // uninitialized filter
+    if ( mFilterComboBoxes.count() > 0
```

Comment:
`mFilterComboBoxes` is a `QList`? Use `isEmpty()` instead (it's more performant).

Reply:
Of course, silly me...

Reply:
#pedantic

I don't think it's actually faster (I looked into this once), but it *is* much easier to understand the logic behind the condition by using empty() instead.

Other dumb #pedantic thing: if you use empty() instead of isEmpty(), it'll be easier for us to change to std containers in future...

Reply:
Totally unrelated comment:

In this case it's `QgsVectorLayer::empty()` and not `QgsVectorLayer::isEmpty()` which I'm going to add soon.

Reply:
QgsFeatureSource::empty()? (pretty please :angel: )

Reply:
Hehe, I'll have a look but can't guarantee because I'm not sure it's trivial to know for all feature sources.

Either

 * Introduce `QgsFeatureSource::capabilities()` (`KnowsIfItIsEmpty`)
 * Introduce `QgsFeatureSource::canDetermineEmpty()`
 * The default implementation consumes the whole source until it either reaches the end or finds an entry. Potentially expensive and might leave resources hanging.
 * If in doubt define if `false` or `true` should be returned
 * Use the unavailable `std::optional<bool>`

Reply:
There's already QgsFeatureSource::featureCount - that could be used as a fallback?

Reply:
If that works, `empty()` works as well :+1:

Reply:
> There's already QgsFeatureSource::featureCount - that could be used as a fallback?

Looking at this in detail, the question remains open. What to do when featureCount() returns -1

Reply:
Return an enum? Empty/NotEmpty/Unknown?

Reply:
I'll intuitively always write `if ( layer->empty() ) ` because I'm used to a method called `empty()` to return a boolean.
And depending on the feature source, this is a safe thing to do (e.g. if I know it's a pg layer).

Maybe NotEmpty = 0, Empty = 1, Unknown = 2 would work.

Reply:
I don't like that, because it means unknown state layers will be skipped by a `if ( layer->empty() )` check. Maybe we should just treat -1 as not empty, since that's going to be the far more common case?

Reply:
In this case we'll at least want a different method that let's us query the unknown state which the dev is free (not to) use.